src/test/libcephfs/snapdiff.cc

```diff
@@ -0,0 +1,1669 @@
+      r = -ENOTEMPTY;
+    }
+    return r;
+  }
+  int for_each_readdir_snapdiff(const char* relpath,
+    const char* snap1,
+    const char* snap2,
+    std::function<bool(const dirent*, uint64_t)> fn)
+  {
+    auto s1 = make_snap_name(snap1);
+    auto s2 = make_snap_name(snap2);
+    ceph_snapdiff_info info;
+    ceph_snapdiff_entry_t res_entry;
+    int r = ceph_open_snapdiff(cmount,
+                               dir_path,
+                               relpath,
+                               s1.c_str(),
+                               s2.c_str(),
+                               &info);
+    if (r != 0) {
+      std::cerr << " Failed to open snapdiff, ret:" << r << std::endl;
+      return r;
+    }
+    while (0 < (r = ceph_readdir_snapdiff(&info,
+                                          &res_entry))) {
+      if (strcmp(res_entry.dir_entry.d_name, ".") == 0 ||
+        strcmp(res_entry.dir_entry.d_name, "..") == 0) {
+        continue;
+      }
+      if (!fn(&res_entry.dir_entry, res_entry.snapid)) {
+        r = -EINTR;
+        break;
+      }
+    }
+    ceph_assert(0 == ceph_close_snapdiff(&info));
+    if (r != 0) {
+      std::cerr << " Failed to readdir snapdiff, ret:" << r
+                << " " << relpath << ", " << snap1 << " vs. " << snap2
+                << std::endl;
+    }
+    return r;
+  }
+  int readdir_snapdiff_and_compare(const char* relpath,
+    const char* snap1,
+    const char* snap2,
+    const vector<pair<string, uint64_t>>& expected0)
+  {
+    vector<pair<string, uint64_t>> expected(expected0);
+    auto end = expected.end();
+    int r = for_each_readdir_snapdiff(relpath, snap1, snap2,
+      [&](const dirent* dire, uint64_t snapid) {
+
+        pair<string, uint64_t> p = std::make_pair(dire->d_name, snapid);
+        auto it = std::find(expected.begin(), end, p);
+        if (it == end) {
+          std::cerr << "readdir_snapdiff_and_compare error: unexpected name:"
+            << dire->d_name << "/" << snapid << std::endl;
+          return false;
+        }
+        expected.erase(it);
+        return true;
+      });
+    if (r == 0 && !expected.empty()) {
+      std::cerr << __func__ << " error: left entries:" << std::endl;
+      for (auto& e : expected) {
+        std::cerr << e.first << "/" << e.second << std::endl;
+      }
+      std::cerr << __func__ << " ************" << std::endl;
+      r = -ENOTEMPTY;
+    }
+    return r;
+  }
+
+  int mkdir(const char* relpath)
+  {
+    auto path = make_file_path(relpath);
+    return ceph_mkdir(cmount, path.c_str(), 0777);
+  }
+  int rmdir(const char* relpath)
+  {
+    auto path = make_file_path(relpath);
+    return ceph_rmdir(cmount, path.c_str());
+  }
+  int purge_dir(const char* relpath0, bool inclusive = true)
+  {
+    int r =
+      for_each_readdir(relpath0,
+        [&] (const dirent* dire) {
+          string relpath = concat_path(relpath0, dire->d_name);
+          if (dire->d_type == DT_REG) {
```

Comment:
The ``dirent`` structure does not include ``d_type`` on Windows. Can we use ``ceph_readdirplus_r`` instead?

Here's a diff:

```
diff --git a/src/include/compat.h b/src/include/compat.h
index 517e9d4fcd5..b19662e4d6d 100644
--- a/src/include/compat.h
+++ b/src/include/compat.h
@@ -251,9 +251,6 @@ typedef unsigned int uint;

 typedef _sigset_t sigset_t;

-typedef unsigned int uid_t;
-typedef unsigned int gid_t;
-
 typedef unsigned int blksize_t;
 typedef unsigned __int64 blkcnt_t;
 typedef unsigned short nlink_t;
diff --git a/src/include/win32/fs_compat.h b/src/include/win32/fs_compat.h
index 2fce1b72e85..635ae8bb6ab 100644
--- a/src/include/win32/fs_compat.h
+++ b/src/include/win32/fs_compat.h
@@ -40,3 +40,6 @@

 #define XATTR_CREATE  1
 #define XATTR_REPLACE 2
+
+typedef unsigned int uid_t;
+typedef unsigned int gid_t;
diff --git a/src/test/libcephfs/snapdiff.cc b/src/test/libcephfs/snapdiff.cc
index 72e7d77ec04..ec87abc0f68 100644
--- a/src/test/libcephfs/snapdiff.cc
+++ b/src/test/libcephfs/snapdiff.cc
@@ -148,7 +148,7 @@ public:
   }

   int for_each_readdir(const char* relpath,
-    std::function<bool(const dirent* dire)> fn)
+    std::function<bool(const dirent*, const struct ceph_statx*)> fn)
   {
     auto subdir_path = make_file_path(relpath);
     struct ceph_dir_result* ls_dir;
@@ -156,17 +156,34 @@ public:
     if (r != 0) {
       return r;
     }
-    struct dirent* result;
-    while( nullptr != (result = ceph_readdir(cmount, ls_dir))) {
-      if (strcmp(result->d_name, ".") == 0 ||
-          strcmp(result->d_name, "..") == 0) {
+
+    while (1) {
+      struct dirent result;
+      struct ceph_statx stx;
+
+      r = ceph_readdirplus_r(
+        cmount, ls_dir, &result, &stx, CEPH_STATX_BASIC_STATS,
+        0,
+        NULL);
+      if (!r)
+        break;
+      if (r < 0) {
+        std::cerr << "ceph_readdirplus_r failed, error: "
+                  << r << std::endl;
+        return r;
+      }
+
+      if (strcmp(result.d_name, ".") == 0 ||
+          strcmp(result.d_name, "..") == 0) {
         continue;
       }
-      if (!fn(result)) {
+      if (!fn(&result, &stx)) {
         r = -EINTR;
         break;
       }
     }
     ceph_assert(0 == ceph_closedir(cmount, ls_dir));
     return r;
   }
@@ -176,7 +193,7 @@ public:
     vector<string> expected(expected0);
     auto end = expected.end();
     int r = for_each_readdir(relpath,
-      [&](const dirent* dire) {
+      [&](const dirent* dire, const struct ceph_statx* stx) {

         std::string name(dire->d_name);
         auto it = std::find(expected.begin(), end, name);
@@ -281,13 +298,13 @@ public:
   {
     int r =
       for_each_readdir(relpath0,
-        [&] (const dirent* dire) {
+        [&] (const dirent* dire, const struct ceph_statx* stx) {
           string relpath = concat_path(relpath0, dire->d_name);
-          if (dire->d_type == DT_REG) {
-            unlink(relpath.c_str());
-          } else if (dire->d_type == DT_DIR) {
+          if (S_ISDIR(stx->stx_mode)) {
             purge_dir(relpath.c_str());
             rmdir(relpath.c_str());
+          } else {
+            unlink(relpath.c_str());
           }
           return true;
         });
```

Reply:
Thanks @petrutlucian94 for noticing the windows build failure. I _really_ need to start paying more attention to those (at least the builds).

Reply:
Indeed, thanks a lot @petrutlucian94 for the diffs and suggestions. Now it looks fine for Win build as well!